sdk/eventhub/azure-eventhub/azure/eventhub/aio/_consumer_async.py

```diff
@@ -171,54 +175,67 @@ def _next_message_in_buffer(self):
-            retried_times = 0
-            has_not_fetched_once = False
-            while retried_times <= max_retries:
-                try:
-                    await self._open()
-                    await cast(ReceiveClientAsync, self._handler).do_work_async(batch=self._prefetch)
-                    break
-                except asyncio.CancelledError:  # pylint: disable=try-except-raise
-                    raise
-                except Exception as exception:  # pylint: disable=broad-except
-                    if (
+        retried_times = 0
+        while retried_times <= max_retries:
+            try:
+                await self._open()
+                await cast(ReceiveClientAsync, self._handler).do_work_async(batch=self._prefetch)
+            except asyncio.CancelledError:  # pylint: disable=try-except-raise
+                raise
+            except Exception as exception:  # pylint: disable=broad-except
+                if (
                         isinstance(exception, error.AMQPLinkError)
                         and exception.condition == error.ErrorCondition.LinkStolen  # pylint: disable=no-member
-                    ):
-                        raise await self._handle_exception(exception)
-                    if not self.running:  # exit by close
-                        return
-                    if self._last_received_event:
-                        self._offset = self._last_received_event.offset
-                    last_exception = await self._handle_exception(exception)
-                    retried_times += 1
-                    if retried_times > max_retries:
-                        _LOGGER.info(
-                            "%r operation has exhausted retry. Last exception: %r.",
-                            self._name,
-                            last_exception,
-                        )
-                        raise last_exception
-
-        if self._message_buffer:
-            while self._message_buffer:
-                if batch:
-                    events_for_callback = []  # type: List[EventData]
-                    for _ in range(min(max_batch_size, len(self._message_buffer))):
-                        events_for_callback.append(self._next_message_in_buffer())
-                    await self._on_event_received(events_for_callback)
-                else:
-                    await self._on_event_received(self._next_message_in_buffer())
-        elif max_wait_time:
-            if batch:
-                await self._on_event_received([])
-            else:
-                await self._on_event_received(None)
+                ):
+                    raise await self._handle_exception(exception)
+                if not self.running:  # exit by close
+                    return
+                if self._last_received_event:
+                    self._offset = self._last_received_event.offset
+                last_exception = await self._handle_exception(exception)
```

Comment:
Do we think it could be a problem that these are now handled from a separate coroutine to the on-event callbacks? Could it lead someone to have to start locking things?

Reply:
I don't think so..

I think for user it's always there's one coroutine calls on_event callbacks.

The internal buffered queue is for decoupling receiving message + calling callbacks.

how we receive messages and dumps into a local queue is an implementation detail.
It happened that previously we do everything in a single coroutine, but now we split out the receiving part. however, the callback part for a user doesn't change.